miniapps/tools/least-squares.cpp

```diff
@@ -0,0 +1,1720 @@
+      InvTr inv_tr(&e_trans);
+      inv_tr.SetPrintLevel(newton.print_level);
+      inv_tr.SetMaxIter(newton.max_iter);
+      inv_tr.SetPhysicalRelTol(newton.rtol);
+      inv_tr.SetSolverType(InvTr::SolverType::Newton);
+      inv_tr.SetInitialGuessType(InvTr::InitGuessType::ClosestPhysNode);
+      MFEM_VERIFY(inv_tr.Transform(phys_neighbor_ip,
+                                   ngh_ip) != InvTr::TransformResult::Unknown, "");
+
+      neighbor_trans.SetIntPoint(&ngh_ip);
+      e_trans.SetIntPoint(&ip);
+
+      // Compute shape functions on e_fe
+      neighbor_fe.CalcPhysShape(neighbor_trans, neighbor_shape);
+      e_fe.CalcPhysShape(e_trans, e_shape);
+
+      e_shape *= e_trans.Weight() * ip.weight;
+      AddMultVWt(e_shape, neighbor_shape, elmat);
+   }
+}
+
+/// @brief Out-of-element symmetric mass matrix (misnomer).
+/// Shape functions come from neighboring element.
+void AsymmetricMassIntegrator::AsymmetricElementMatrix(const FiniteElement
+                                                       &neighbor_fe,
+                                                       ElementTransformation &neighbor_trans,
+                                                       ElementTransformation &e_trans,
+                                                       DenseMatrix &elmat,
+                                                       IterativeSolverParams &newton)
+{
+   neighbor_ndof = neighbor_fe.GetDof();
+
+   neighbor_shape.SetSize(neighbor_ndof);
+   elmat.SetSize(neighbor_ndof);
+   elmat = 0.0;
+
+   const int int_order = 2*neighbor_fe.GetOrder() + e_trans.OrderW();
+   const IntegrationRule &ir = IntRules.Get(neighbor_fe.GetGeomType(), int_order);
+
+   neighbor_trans.Transform(ir, phys_neighbor_pts);
+
+   for (int i = 0; i < ir.GetNPoints(); i++)
+   {
+      const IntegrationPoint &ip = ir.IntPoint(i);
+      phys_neighbor_pts.GetColumn(i, phys_neighbor_ip);
+
+      // Pullback physical neighboring integration points to
+      // "extended" reference element under e_trans
+      using InvTr = mfem::InverseElementTransformation;
+      InverseElementTransformation inv_tr(&e_trans);
+      inv_tr.SetPrintLevel(newton.print_level);
+      inv_tr.SetMaxIter(newton.max_iter);
+      inv_tr.SetPhysicalRelTol(newton.rtol);
+      inv_tr.SetSolverType(InvTr::SolverType::Newton);
+      inv_tr.SetInitialGuessType(InvTr::InitGuessType::ClosestPhysNode);
+      MFEM_VERIFY(inv_tr.Transform(phys_neighbor_ip,
+                                   ngh_ip) != InvTr::TransformResult::Unknown, "");
+
+      neighbor_trans.SetIntPoint(&ngh_ip);
+      e_trans.SetIntPoint(&ip);
+
+      neighbor_fe.CalcPhysShape(neighbor_trans, neighbor_shape);
+      AddMult_a_VVt(e_trans.Weight()*ip.weight, neighbor_shape, elmat);
+   }
+}
+
+
+/** Integrator of the DPG form $\langle v, \lbrace w \rbrace \rangle$ where
+ *  $\lbrace \cdot \rbrace$ represents the average of the traces of the argument.
+ */
+class TraceAverageIntegrator : public BilinearFormIntegrator
```

Comment:
Similar to a comment about the `AsymmetricMassIntegrator`, does this class need to inherit from `BilinearFormIntegrator`?  More specifically, is a `TraceAverageIntegrator` ever used as a `BilinearFormIntegrator`?  If not, please remove this inheritance.  For what it's worth, this is known as preferring composition ("has a") over inheritance ("is a").